sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/AmqpMetricsProvider.java

```diff
@@ -0,0 +1,391 @@
+        Object enqueuedTimeDate = properties != null ? properties.get(ENQUEUED_TIME_ANNOTATION) : null;
+        if (enqueuedTimeDate instanceof Date) {
+            Instant enqueuedTime = ((Date) enqueuedTimeDate).toInstant();
+            long deltaMs = Instant.now().toEpochMilli() - enqueuedTime.toEpochMilli();
+            if (deltaMs < 0) {
+                deltaMs = 0;
+            }
+            receivedLag.record(deltaMs / 1000d, commonAttributes, Context.NONE);
+        } else {
+            LOGGER.verbose("Received message has unexpected `x-opt-enqueued-time` annotation value - `{}`. Ignoring it.", enqueuedTimeDate);
+        }
+    }
+
+    /**
+     * Records that credits were added to link
+     */
+    public void recordAddCredits(int credits) {
+        if (isEnabled && addCredits.isEnabled()) {
+            addCredits.add(credits, commonAttributes, Context.NONE);
+        }
+    }
+
+    /**
+     * Records link error. Noop if condition is null (no error).
+     */
+    public void recordHandlerError(ErrorSource source, ErrorCondition condition) {
+        if (isEnabled && condition != null && condition.getCondition() != null) {
+            TelemetryAttributes attributes = amqpErrorAttributeCache.getOrCreate(condition.getCondition().toString());
+            switch (source) {
+                case LINK:
+                    if (linkErrors.isEnabled()) {
+                        linkErrors.add(1, attributes, Context.NONE);
+                    }
+                    break;
+                case SESSION:
+                    if (sessionErrors.isEnabled()) {
+                        sessionErrors.add(1, attributes, Context.NONE);
+                    }
+                    break;
+                case TRANSPORT:
+                    if (transportErrors.isEnabled()) {
+                        transportErrors.add(1, attributes, Context.NONE);
+                    }
+                    break;
+                default:
+                    LOGGER.verbose("Unexpected error source: {}", source);
+            }
+        }
+    }
+
+    private static String deliveryStateToLowerCaseString(DeliveryState.DeliveryStateType state) {
+        if (state == null) {
+            return "error";
+        }
+
+        switch (state) {
+            case Accepted:
+                return "accepted";
+            case Declared:
+                return "declared";
+            case Modified:
+                return "modified";
+            case Received:
+                return "received";
+            case Rejected:
+                return "rejected";
+            case Released:
+                return "released";
+            case Transactional:
+                return "transactional";
+            default:
+                return "unknown";
+        }
+    }
+
+    private static String deliveryStateToLowerCaseString(AmqpResponseCode response) {
+        if (response == null) {
+            return "error";
+        }
+
+        switch (response) {
```

Comment:
**extremely** minor nit: different amounts of newlines between case statements here.